### PR TITLE
Add wasm target

### PR DESCRIPTION
**Pull request by SuperCaliMan**
Hi, thanks for getting into this project!

Checklist for your PR:
- [X] Check if there's any open issue
- [X] Please file your request against the _develop_ branch

# Description of your changes
Hi i implemented wasm target, to import and use this library in a wasm application. 

# How has this been tested?
I test it with OpenID  [link](https://openidconnect.net/)

# Is this a (API-) breaking change?
No

### CODE REVIEW

**Comment by kalinjul** [comment]
Hi, this only adds wasm target without providing implementation, right?
Why does this need ktor eap?

**Comment by SuperCaliMan** [comment]
> Hi, this only adds wasm target without providing implementation, right? Why does this need ktor eap?

Yes. Ktor support wasm target only in eap

**Comment by ralfstuckert** [comment]
Any news on this PR? This feature would be a great benefit

**Comment by kalinjul** [comment]
Ok so there are some things in this PR I'd like to discuss, i'll put my questions in a review.
But also I'd like to wait for a stable ktor release supporting wasm before requiring a beta or EAP in this library.

File: oidc-appsupport/src/wasmJsMain/kotlin/org/publicvalue/multiplatform/oidc/appsupport/PlatformOidcCodeAuthFlow.wasmJs.kt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                state = resultParameters["state"]
+            )
+        )
+
+    }
+
+    suspend fun waitForRedirectUri(loginWindow: Window?): Parameters? {
+        while (true) {
+            var parameters: Parameters? = null
+            jsCatch {
```
Comment (kalinjul) [inline oidc-appsupport/src/wasmJsMain/kotlin/org/publicvalue/multiplatform/oidc/appsupport/PlatformOidcCodeAuthFlow.wasmJs.kt:38]: Why do we need an additional jsCatch here?
Can't we just catch Throwable or dynamic as mentioned here https://discuss.kotlinlang.org/t/how-to-catch-non-kotlin-exception/2404/14

File: oidc-core/src/commonMain/kotlin/org/publicvalue/multiplatform/oidc/types/Jwt.kt [hunk 2]
```diff
@@ -133,19 +133,19 @@ value class JwtClaims(
...
+        acr = claims["acr"] as? String?,
         amr = claims["amr"]?.parseListOrString(),
-        azp = claims["azp"] as String?,
-        alg = claims["alg"] as String?,
-        kid = claims["kid"] as String?,
-        at_hash = claims["at_hash"] as String?,
+        azp = claims["azp"] as? String?,
+        alg = claims["alg"] as? String?,
+        kid = claims["kid"] as? String?,
+        at_hash = claims["at_hash"] as? String?,
```
[the remark this answers is not in the document]
Comment (SuperCaliMan) [reply oidc-core/src/commonMain/kotlin/org/publicvalue/multiplatform/oidc/types/Jwt.kt:148]: "as?" is required for a safe-cast. You're right, iss, sub etc should be non-nullable in accordance with [RFC 7519](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.1). I will fix it

File: oidc-appsupport/src/wasmJsMain/kotlin/org/publicvalue/multiplatform/oidc/appsupport/PlatformOidcCodeAuthFlow.wasmJs.kt (hunk 1 above)
Comment (SuperCaliMan) [reply oidc-appsupport/src/wasmJsMain/kotlin/org/publicvalue/multiplatform/oidc/appsupport/PlatformOidcCodeAuthFlow.wasmJs.kt:38]: We can only catch Throwable, I'm trying to catch dynamic but i get an error: "Unsupported [Dynamic types are not supported in this context". I think Throwable is enough

**Comment by maicol07** [comment]
Hi, any news on this? Ktor 3 is now in RC and it will soon be stable

**Comment by SuperCaliMan** [comment]
> Hi, any news on this? Ktor 3 is now in RC and it will soon be stable

Hi, I will update this PR when Ktor 3 is released in the stable channel.

**Comment by the-best-is-best** [comment]
please uodate ktor version 3 in rc version becouse it is make issue with another packages in compose multiplatform koin

**Comment by SuperCaliMan** [comment]
> please uodate ktor version 3 in rc version becouse it is make issue with another packages in compose multiplatform

Ktor 3.0.0-rc-1 require to update compose version

**Comment by DatL4g** [comment]
Ktor 3.0.0 https://github.com/ktorio/ktor/releases/tag/3.0.0 is stable now

File: oidc-core/src/commonMain/kotlin/org/publicvalue/multiplatform/oidc/DefaultOpenIdConnectClient.kt [hunk 3]
```diff
@@ -166,14 +168,14 @@ class DefaultOpenIdConnectClient(
             config.clientSecret?.let { append("client_secret", it) }
             if (config.codeChallengeMethod != CodeChallengeMethod.off) { append("code_verifier", authCodeRequest.pkce.codeVerifier) }
         }
-        val request = runBlocking { // there is no suspending happening here
+        val request = scope.async { // there is no suspending happening here
```
[the remark this answers is not in the document]
Comment (kalinjul) [reply oidc-core/src/commonMain/kotlin/org/publicvalue/multiplatform/oidc/DefaultOpenIdConnectClient.kt:171]: Could you elaborate on this? Are there cases where the httpClient.prepareForm function suspends? Why would we want a SupervisorJob?

[the remark this answers is not in the document]
Comment (DatL4g) [reply oidc-core/src/commonMain/kotlin/org/publicvalue/multiplatform/oidc/DefaultOpenIdConnectClient.kt:171]: @kalinjul for clarification, JS targets do not support runBlocking
https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/run-blocking.html

Supported targets are JVm and Native only

**Comment by SuperCaliMan** [comment]
Hi, I have just upgraded ktor to stable version 3.0.1.

**Comment by SuperCaliMan** [comment]
Hi, i'm still waiting @kalinjul review

**Comment by kalinjul** [comment]
This was closed automatically because i deleted the develop branch, could you change to the main branch?

We'll need this rebased, as main has been updated to ktor 3 as well.
Also, there are still TODOs in the code, so i guess it's not ready yet.

**Comment by kalinjul** [comment]
@SuperCaliMan can you reopen this or make a new pr to main branch?

**Pull request closed without merging**